### PR TITLE
Creating index template for Kibana to configure index replicas

### PR DESCRIPTION
**Pull request by ewolinetz**
Addresses https://github.com/openshift/origin-aggregated-logging/issues/1315

### CODE REVIEW

File: elasticsearch/index_templates/common.settings.kibana.template.json [hunk 1]
```diff
@@ -0,0 +1,8 @@
+{
+  "order": 0,
+  "settings": {
+    "index.number_of_replicas": $REPLICA_SHARDS,
+    "index.number_of_shards": PRIMARY_SHARDS
```
Comment (lukas-vlcek) [inline elasticsearch/index_templates/common.settings.kibana.template.json]: Missing `$` ? (`$PRIMARY_SHARDS`)

**Comment by lukas-vlcek** [comment]
Left one comment.
Apart from this LGTM

**Comment by ewolinetz** [comment]
Squashed commits

**Comment by ewolinetz** [comment]
/hold
looking on my local deployment I don't see that Kibana has the correct replica count even though our template is there with replicas 0

```
$ oc exec -c elasticsearch logging-es-data-master-n95jnq3h-1-tvqg4 -- es_util --query=_template/common.settings.kibana.template.json

{"common.settings.kibana.template.json":{"order":0,"template":".kibana*","settings":{"index":{"number_of_shards":"1","number_of_replicas":"0"}},"mappings":{},"aliases":{}}}


health status index                                                           uuid                   pri rep docs.count docs.deleted store.size pri.store.size
yellow open   .kibana                                                         o0U2BGXbTt6U9LGLGaFJKw   1   1          1            0      3.2kb          3.2kb
```

**Comment by lukas-vlcek** [comment]
@ewolinetz let's check how exactly Kibana creates its indices. Can we locate particular line in the source code?

**Comment by ewolinetz** [comment]
@lukas-vlcek 

From the ES logs:
```
...
[2018-08-28T21:58:44,153][WARN ][c.f.s.c.PrivilegesEvaluator] .kibana does not exist in cluster metadata
[2018-08-28T21:58:44,308][INFO ][o.e.c.m.MetaDataCreateIndexService] [logging-es-data-master-2kezn13a] [.kibana] creating index, cause [api], templates [], shards [1]/[1], mappings [_de
fault_, index-pattern, server, visualization, search, timelion-sheet, config, dashboard, url]
Elasticsearch Version: 5.6.10
Search Guard Version: <unknown>
Reload config on all nodes
Auto-expand replicas disabled
/usr/share/elasticsearch/init
[2018-08-28 21:58:46,797][INFO ][container.run            ] Updating replica count to 0
/etc/elasticsearch /usr/share/elasticsearch/init
Search Guard Admin v5
Will connect to localhost:9300 ... done
ERROR StatusLogger No log4j2 configuration file found. Using default configuration: logging only errors to the console. Set system property 'log4j2.debug' to show Log4j2 internal initia
lization logging.
[2018-08-28T21:58:49,674][WARN ][c.f.s.a.BackendRegistry  ] Authentication finally failed for null
Elasticsearch Version: 5.6.10
Search Guard Version: <unknown>
[2018-08-28T21:58:54,593][INFO ][o.e.c.m.MetaDataUpdateSettingsService] [logging-es-data-master-2kezn13a] updating number_of_replicas to [0] for indices [.searchguard]
Reload config on all nodes
Update number of replicas to 0 with result: true
/usr/share/elasticsearch/init
[2018-08-28 21:58:55,136][INFO ][container.run            ] Setting filters to allocate shard to this node only
{"acknowledged":true}[2018-08-28 21:58:55,428][INFO ][container.run            ] Adding index templates
[2018-08-28 21:58:55,679][INFO ][container.run            ] Create index template 'com.redhat.viaq-openshift-operations.template.json'
{"acknowledged":true}[2018-08-28 21:58:56,560][INFO ][container.run            ] Create index template 'com.redhat.viaq-openshift-orphaned.template.json'
{"acknowledged":true}[2018-08-28 21:58:57,508][INFO ][container.run            ] Create index template 'com.redhat.viaq-openshift-project.template.json'
{"acknowledged":true}[2018-08-28 21:58:58,129][INFO ][container.run            ] Create index template 'common.settings.kibana.template.json'
{"acknowledged":true}[2018-08-28 21:58:58,830][INFO ][container.run            ] Create index template 'common.settings.operations.orphaned.json'
{"acknowledged":true}[2018-08-28 21:58:59,392][INFO ][container.run            ] Create index template 'common.settings.operations.template.json'
{"acknowledged":true}[2018-08-28 21:59:00,078][INFO ][container.run            ] Create index template 'common.settings.project.template.json'
{"acknowledged":true}[2018-08-28 21:59:00,858][INFO ][container.run            ] Create index template 'org.ovirt.viaq-collectd.template.json'
{"acknowledged":true}[2018-08-28 21:59:01,367][INFO ][container.run            ] Finished adding index templates
...
```
From `/usr/share/kibana/src/core_plugins/elasticsearch/lib/create_kibana_index.js`:
```js
return callWithInternalUser('indices.create', {
    index: index,
    body: {
      settings: {
        number_of_shards: 1,
        'index.mapper.dynamic': false
      },
      mappings
    }
  }).catch(handleError('Unable to create Kibana index "<%= kibana.index %>"')).then(function () {
    return callWithInternalUser('cluster.health', {
      waitForStatus: 'yellow',
      index: index
    }).catch(handleError('Waiting for Kibana index "<%= kibana.index %>" to come online failed.'));
  });
```

**Comment by anpingli** [comment]
@ewolinetz   What will happen if we delete the .kibana index manually.  Is the new .kibana index replicas=0?

**Comment by lukas-vlcek** [comment]
@ewolinetz the `.kibana` index is create with `shards [1]/[1]` (1 primary, 1 replica shards) way before we are putting out index templates into the cluster. I do not know where the number of replicas is configured in this case, however, if we just want to change the number of replicas for the testing case then we can simply change it on the fly using the [index settings update API](https://www.elastic.co/guide/en/elasticsearch/reference/5.6/indices-update-settings.html#indices-update-settings).

**Comment by ewolinetz** [comment]
@lukas-vlcek unfortunately when i've tried that before the request is redirected to a kibana user index instead of changing the `.kibana` index.

@anpingli I'll try that and see. Hopefully it doesn't redirect it to a kibana user index as well...

**Comment by lukas-vlcek** [comment]
@ewolinetz even when using admin certs?

**Comment by jcantrill** [comment]
The multitenant code will attempt to modify anything that has Kibana in the request and a bearer token is provided.  This is the source of a known issue where we are unable to delete index patterns.  Maybe we should consider allowing admin users who provide a special header to take a request as is with no modifications.  I would expect the admin certs to already behave this way but maybe there is something we are modifying

**Comment by lukas-vlcek** [comment]
FYI, the `number_of_replicas: 1` value comes from the default values. See <https://github.com/elastic/elasticsearch/blob/master/docs/reference/indices/create-index.asciidoc#index-settings>. Seems other people were [running into YELLOW cluster](https://github.com/elastic/kibana/issues/3877#issuecomment-345010371) issue in case of a single node cluster too.

May be we can accept the YELLOW cluster state when running particular test?

**Comment by ewolinetz** [comment]
I can delete the indices, and then I see them recreated correctly with 0 replicas.
```
[2018-08-29T15:48:02,816][INFO ][o.e.c.m.MetaDataDeleteIndexService] [logging-es-data-master-aw89bch5] [.kibana/XDC_dKYWSTyWCDJ0_9wISw] deleting index
[2018-08-29T15:48:09,983][WARN ][c.f.s.c.PrivilegesEvaluator] .kibana does not exist in cluster metadata
[2018-08-29T15:48:09,990][INFO ][o.e.c.m.MetaDataCreateIndexService] [logging-es-data-master-aw89bch5] [.kibana] creating index, cause [api], templates [common.settings.kibana.template.json], shards [1]/[0], mappings [_default_, index-pattern, server, visualization, search, timelion-sheet, config, dashboard, url]
[2018-08-29T15:48:10,015][INFO ][o.e.c.r.a.AllocationService] [logging-es-data-master-aw89bch5] Cluster health status changed from [YELLOW] to [GREEN] (reason: [shards started [[.kibana][0]] ...]).
```

**Comment by ewolinetz** [comment]
I think we are running into an issue where we are not ready (the index templates are still being seeded) and then ES is somehow allowing traffic through.

**Comment by ewolinetz** [comment]
Confirmed that https://github.com/openshift/origin-aggregated-logging/pull/1326 fixes the issue of the index being created before the index template is seeded

**Pull request merged**